pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -1975,8 +1981,17 @@ private void failPendingBatchMessages(PulsarClientException ex) {
             if (isBatchMessagingEnabled()) {
                 batchMessageAndSend();
             }
-            lastSendFuture = this.lastSendFuture;
+            if (lastSendFutureResponse) {
+                lastSendFuture = this.lastSendFutureEmpty;
+            } else {
+                lastSendFuture = this.lastSendFuture;
+                lastSendFuture.exceptionally(ignored -> {
+                    lastSendFutureResponse = true;
+                    return null;
+                });
+            }
         }
+
         return lastSendFuture.thenApply(ignored -> null);
```

Comment:
Since the intention seems to be to ignore an exception that is delivered once, it would be better to make that more explicit. It would be a breaking change if calling flushAsync wouldn't deliver a possible exception to the caller at all.

To prevent race conditions this would have to be implemented in a different way. The high level solution would be to create a class that contains 2 fields: a completable future and a AtomicBoolean field (or volate. The "lastSendFuture" field should reference this wrapper and ProducerImpl should delegate actions to it. The wrapper would contain behavior that makes it return the exception only for the "first call". Technically this could be implemented using CompletableFuture.handle .  It's possible to re-throw the exception in handle when wrapped with CompletionException. That would be done only the "first call". AtomicBoolean's compareAndSet could be used to ensure that the exception is thrown only on the "first call". ("first call" isn't really a call in this case, but I found it easier to explain it that way)

Reply:
Thanks for your review. I added a `LastSendFutureWrapper` to wrap `lastSendFuture`. Is this implementation correct? Can you help take a look? Thanks.